swim/join_sender.go

```diff
@@ -73,6 +77,10 @@ type joinSender struct {
 	node    *Node
 	timeout time.Duration
 
+	// bootstrapHostsMap is a map of unique hosts each containing a slice of
+	// the instances (hostsports) on that particular host.
+	bootstrapHostsMap map[string][]string
```

Comment:
OK. Your perspective is clearer. I was coming from the perspective that the join sender and its fan-out is a separate concern from how the join sender selects nodes from the bootstrap list. The process of grouping by host would be encapsulated within the "Bootstrap Selector-type-of-abstraction."

I think it would be fine for the DiscoverProvider to require that some context (like the identity of the node calling it) would be passed to it. If you want to leave the selection concern out of the DiscoverProvider for now, I'd be OK with that.